core/src/backend/audio/decoders.rs

```diff
@@ -600,3 +604,139 @@ impl Iterator for AdpcmSubstreamDecoder {
+                .with_channel_layout(layout)
+                .with_sample_rate(sample_rate);
+
+            let decoder = SymphoniaAacDecoder::try_new(&codec_params, &Default::default())?;
+
+            Ok(Self {
+                tag_reader,
+                decoder,
+                sample_buf: audio::SampleBuffer::new(
+                    0,
+                    audio::SignalSpec::new(sample_rate, layout.into_channels()),
+                ),
+                cur_sample: 0,
+                stream_ended: false,
+            })
+        }
+    }
+
+    impl Decoder for AacSubstreamDecoder {
+        fn num_channels(&self) -> u8 {
+            self.decoder.last_decoded().spec().channels.count() as u8
+        }
+        fn sample_rate(&self) -> u16 {
+            self.decoder.last_decoded().spec().rate as u16
+        }
+    }
+
+    impl Iterator for AacSubstreamDecoder {
+        type Item = [i16; 2];
+
+        #[inline]
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.cur_sample >= self.sample_buf.len() {
+                if !self.stream_ended {
+                    self.stream_ended = true;
+                    self.cur_sample = 0;
+                    for chunk in Iterator::by_ref(&mut self.tag_reader) {
+                        match chunk.data()[0] {
+                            0 => {
+                                let mut codec_params = CodecParameters::new();
+                                codec_params.for_codec(core::codecs::CODEC_TYPE_AAC);
+                                codec_params.with_extra_data(chunk.data()[1..].into());
+
+                                self.decoder = SymphoniaAacDecoder::try_new(
+                                    &codec_params,
+                                    &Default::default(),
+                                )
+                                .unwrap();
+
+                                let spec = *(self.decoder.last_decoded().spec());
+                                self.sample_buf = audio::SampleBuffer::new(0, spec);
+                            }
+                            1 => {
+                                let packet = Packet::new_from_slice(0, 0, 0, &chunk.data()[1..]);
+                                match self.decoder.decode(&packet) {
+                                    Ok(decoded) => {
+                                        if self.sample_buf.capacity() < decoded.capacity() {
+                                            // Ensure our buffer has enough space for the decoded samples.
+                                            self.sample_buf = audio::SampleBuffer::new(
+                                                decoded.capacity() as core::units::Duration,
+                                                *decoded.spec(),
+                                            );
+                                        }
+                                        self.sample_buf.copy_interleaved_ref(decoded);
```

Comment:
There's some code duplication with `mp3`, maybe we should add some utility method or `SampleBuffer` wrapper?

Reply:
Yeah, it would be neat, but:
 - It's symphonia-specific, and only 2 of our decoders are based on symphonia.
 - I believe 2 is just below the threshold of tolerable copy-pasting, and doesn't quite reach the threshold of "an abstraction would be more helpful".
 - There are some subtle differences in the two instances (see `self.stream_ended = false;`).
 - The optional assignment to `self.sample_buf` makes it tricky.

Reply:
Sure, that's perfectly fine!